zcash_client_backend/src/fees.rs

```diff
@@ -0,0 +1,163 @@
+use zcash_primitives::{
+    consensus::{self, BlockHeight},
+    transaction::{
+        components::{
+            amount::{Amount, BalanceError},
+            sapling::builder::{SaplingInput, SaplingOutput},
+            transparent::{builder::TransparentInput, TxOut},
+        },
+        fees::{FeeRule, FixedFeeRule},
+    },
+};
+
+/// A proposed change amount and output pool.
+pub enum ChangeValue {
+    Sapling(Amount),
+}
+
+impl ChangeValue {
+    pub fn value(&self) -> Amount {
+        match self {
+            ChangeValue::Sapling(value) => *value,
+        }
+    }
+}
+
+/// The amount of change and fees required to make a transaction's inputs and
+/// outputs balance under a specific fee rule, as computed by a particular
+/// [`ChangeStrategy`] that is aware of that rule.
+pub struct TransactionBalance {
+    proposed_change: Vec<ChangeValue>,
+    fee_required: Amount,
+}
```

Comment:
I like this approach; it's going to enable note splitting easily. I'm now wondering about note merging; `ChangeStrategy` assumes that all of the provided inputs will be part of the transaction (which makes sense given that it is focused on change strategy), so does this mean to have a variable number of inputs beyond what is necessary, we will need a third layer around `ChangeStrategy` and `FeeRule`? Or would it make more sense to do both note merging and note splitting in the same place? I guess the outer layer could equivalently implement that itself...

Reply:
Your inference is correct: we *do* need a third layer! See #689